cranelift/codegen/src/lib.rs

```diff
@@ -89,7 +89,10 @@ pub mod write;
 
 pub use crate::entity::packed_option;
 pub use crate::machinst::buffer::{MachCallSite, MachReloc, MachSrcLoc, MachStackMap, MachTrap};
-pub use crate::machinst::{CompiledCode, TextSectionBuilder};
+pub use crate::machinst::{
+    CompiledCode, Final, MachBuffer, MachBufferFinalized, MachInst, MachInstEmit, Reg,
+    TextSectionBuilder, Writable,
+};
```

Comment:
I don't think all of these should be exported at the top level. Most people won't use them, so it may be confusing. Would it make sense to make `machinst` public and instead use `pub(crate)` where things are not meant to be exported?

Reply:
I had considered this. But didn't do it mainly for two reasons: (i) there's already a precedent: we expose `CompiledCode` and `TextSectionBuilder` at the top level; so even though there are no other uses cases aside from Winch that will use the other primitives, personally I feel it makes sense to expose the core binary emission concepts at the top level. (ii) If we want to expose them under `machinst`, to be consistent we should probably not expose `CompiledCode` and `TextSectionBuilder` at the top level and move them under `machinst` instead. But this will be a breaking change for anyone relying on those exports.

Since Winch is the only use-case relying on this, I'm happy to change it if/once we have a concrete example of an issue caused by this approach.